oak-blob-cloud-azure/src/main/java/org/apache/jackrabbit/oak/blob/cloud/azure/blobstorage/AzureBlobStoreBackend.java

```diff
@@ -0,0 +1,784 @@
+        }
+        catch (StorageException e) {
+            LOG.info("Error writing blob. identifier={}", key, e);
+            throw new DataStoreException(String.format("Cannot write blob. identifier=%s", key), e);
+        }
+        catch (URISyntaxException | IOException e) {
+            LOG.debug("Error writing blob. identifier={}", key, e);
+            throw new DataStoreException(String.format("Cannot write blob. identifier=%s", key), e);
+        } catch (InterruptedException e) {
+            LOG.debug("Error writing blob. identifier={}", key, e);
+            throw new DataStoreException(String.format("Cannot copy blob. identifier=%s", key), e);
+        } finally {
+            if (null != contextClassLoader) {
+                Thread.currentThread().setContextClassLoader(contextClassLoader);
+            }
+        }
+    }
+
+    public static boolean waitForCopy(CloudBlob blob) throws StorageException, InterruptedException {
+        boolean continueLoop = true;
+        CopyStatus status = CopyStatus.PENDING;
+        while (continueLoop) {
+            blob.downloadAttributes();
+            status = blob.getCopyState().getStatus();
+            continueLoop = status == CopyStatus.PENDING;
+            // Sleep if retry is needed
+            if (continueLoop) {
+                Thread.sleep(500);
+            }
+        }
+        return status == CopyStatus.SUCCESS;
+    }
+
+    @Override
+    public byte[] getOrCreateReferenceKey() throws DataStoreException {
+        try {
+            if (!Strings.isNullOrEmpty(secret)) {
+                return secret.getBytes("UTF-8");
+            }
+            LOG.warn("secret not defined");
+            return super.getOrCreateReferenceKey();
+        } catch (UnsupportedEncodingException e) {
+            throw new DataStoreException(e);
+        }
+    }
+
+    @Override
+    public DataRecord getRecord(DataIdentifier identifier) throws DataStoreException {
+        if (null == identifier) {
+            throw new NullPointerException("identifier");
+        }
+        String key = getKeyName(identifier);
+        long start = System.currentTimeMillis();
+        ClassLoader contextClassLoader = Thread.currentThread().getContextClassLoader();
+        try {
+            Thread.currentThread().setContextClassLoader(getClass().getClassLoader());
+
+            CloudBlockBlob blob = getAzureContainer().getBlockBlobReference(key);
+            if (blob.exists()) {
+                blob.downloadAttributes();
+                AzureBlobStoreDataRecord record = new AzureBlobStoreDataRecord(
+                    this,
+                    connectionString,
+                    containerName,
+                    new DataIdentifier(getIdentifierName(blob.getName())),
+                    blob.getProperties().getLastModified().getTime(),
+                    blob.getProperties().getLength());
```

Comment:
It is intentional since the thread safety of the Azure APIs is not guaranteed (https://github.com/Azure/azure-storage-java/issues/38) . And since the fetching of all dataRecords can be made on a different thread than the one reading one, for now, we're playing the safe card.